src/markov/ddp.jl

```diff
@@ -0,0 +1,472 @@
+
+    # method to pass initial value function (skip the s-wise max)
+    function DPSolveResult(ddp::DiscreteDP, v::Vector)
+        ddpr = new(v, similar(v), 0, similar(v, Int))
+
+        # fill in sigma with proper values
+        compute_greedy!(ddp, ddpr)
+        ddpr
+    end
+end
+
+"""
+Return the `Vector` `max_a vals(s, a)`,  where `vals` is represented as a
+`Matrix` of size `(num_states, num_actions)`.
+"""
+s_wise_max(vals::Matrix) = vec(maximum(vals, 2))
+
+"""
+Populate `out` with  `max_a vals(s, a)`,  where `vals` is represented as a
+`Matrix` of size `(num_states, num_actions)`.
+"""
+s_wise_max!(vals::Matrix, out::Vector) = maximum!(out, vals)
+
+"""
+Populate `out` with  `max_a vals(s, a)`,  where `vals` is represented as a
+`Matrix` of size `(num_states, num_actions)`.
+
+Also fills `out_argmax` with the linear index associated with the indmax in each
+row
+"""
+s_wise_max!(vals::Matrix, out::Vector, out_argmax) =
+    Base.findminmax!(Base.MoreFun(), fill!(out, -Inf) , out_argmax, vals)
+
+#=
+Define Matrix Multiplication between 3-dimensional matrix and a vector
+
+Matrix multiplication over the last dimension of A
+
+=#
+function *{T}(A::Array{T,3}, v::Vector)
+    size(A, 3) ==  size(v, 1) || error("wrong dimensions")
+    out = Array(T, size(A)[1:2])
+
+    # TODO: slicing A[i, j, :] is cache-unfriendly
+    for j=1:size(out, 2), i=1:size(out, 1)
+        out[i, j] = dot(A[i, j, :][:], v)
+    end
+    out
+end
+
+"""
+The Bellman operator, which computes and returns the updated value function Tv
+for a value function v.
+
+##### Parameters
+
+- `ddp::DisreteDP` : Object that contains the model parameters
+- `ddpr::DPSolveResult` : Object that contains result variables
+
+##### Returns
+
+- `ddpr.Tv::Vector` : Updated value function vector
+
+##### Notes
+
+Updates `ddpr.Tv` and `ddpr.sigma` inplace
+
+"""
+function bellman_operator!(ddp::DiscreteDP, ddpr::DPSolveResult)
+    vals = ddp.R + ddp.beta * ddp.Q * ddpr.v
+    s_wise_max!(vals, ddpr.Tv, ddpr.sigma)
+    ddpr.Tv
+end
+
+"""
+The Bellman operator, which computes and returns the updated value function Tv
+for a given value function v.
+
+This function will fill the input `v` with `Tv` and the input `sigma` with the
+corresponding policy rule
+
+##### Parameters
+
+- `ddp::DisreteDP`: The ddp model
+- `v::Vector{T<:AbstractFloat}`: The current guess of the value function
+- `sigma::Vector`: A buffer array to hold the policy function. Initial
+  values not used and will be overwritten
+
+##### Returns
+
+- `Tv::Vector`: Updated value function vector
+- `sigma::Vector{T<:Integer}`: Policy rule
+"""
+function bellman_operator!{T<:AbstractFloat}(ddp::DiscreteDP, v::Vector{T},
+                                             sigma::Vector)
+    vals = ddp.R + ddp.beta * ddp.Q * v
+    s_wise_max!(vals, v, sigma)
+    v, sigma
+end
+
+# method to allow users to pass v as something like v=[1, 1, 1]
+bellman_operator!{T<:Real}(ddp::DiscreteDP, v::Vector{T}, sigma::Vector) =
```

Comment:
No problem about the requests. We can add these methods. Do you worry that the number of API points is getting overwhelming?

And yes, that last method is a bit odd. I'll re-work it. We will lose some efficiency, but it will be the correct behavior.